react/utils/metrics.ts

```diff
@@ -0,0 +1,148 @@
+
+const ANALYTICS_URL = 'https://rc.vtex.com/api/analytics/schemaless-events'
+
+const GRAPHQL_URL = (accountName: string, workspace?: string) => {
+  if (workspace) {
+    return `https://${workspace}--${accountName}.myvtex.com/_v/private/graphql/v1`
+  }
+
+  return `https://${accountName}.myvtex.com/_v/private/graphql/v1`
+}
+
+type Metric = {
+  name: 'b2b-suite-buyerorg-data'
+  kind: 'create-quote-ui-event'
+  description: 'Create Quotation Action - UI'
+  account: string
+}
+
+type QuoteFieldsMetric = {
+  cost_center_id: string
+  cost_center_name: string
+  buy_org_id: string
+  buy_org_name: string
+  member_id: string
+  member_email: string
+  role: string
+  creation_date: string
+  quote_id: string
+  quote_reference_name: string
+  send_to_sales_rep: boolean
+}
+
+export type SessionProfile = {
+  id: { value: string }
+  email: { value: string }
+}
+
+type SessionResponse = {
+  namespaces: {
+    profile: SessionProfile
+    account: {
+      accountName: { value: string }
+    }
+  }
+}
+
+type MetricsParam = {
+  quoteId: string
+  sessionResponse: SessionResponse
+  workspace: string
+  sendToSalesRep: boolean
+}
+
+type QuoteMetric = Metric & { fields: QuoteFieldsMetric }
+
+const fetchMetricsData = async (
+  accountName: string,
+  workspace: string,
+  quoteId: string,
+  userEmail: string
+) => {
+  const query = JSON.stringify({
+    query: `query GetMetricsData($id: String, $email: String!) {
+      getQuote(id: $id) @context(provider: "vtex.b2b-quotes-graphql") {
+        organization
+        organizationName
+        costCenterName
+        referenceName
+        creatorRole
+        creationDate
+      },
+      getUserByEmail(email: $email) @context(provider: "vtex.storefront-permissions") {
+      costId
+    }
+  }`,
+    variables: { id: quoteId, email: userEmail },
+  })
+
+  const { data } = (
+    await axios.post(GRAPHQL_URL(accountName, workspace), query)
+  ).data
+
+  const quoteResult = data.getQuote as Omit<QuoteMetricsData, 'costId'>
+  const costId = data.getUserByEmail?.[0].costId as string
+
+  return {
+    ...quoteResult,
+    costId,
+  }
+}
+
+const buildQuoteMetric = async (
+  metricsParam: MetricsParam
+): Promise<QuoteMetric> => {
+  const { namespaces } = metricsParam.sessionResponse
+  const accountName = namespaces.account.accountName.value
+  const userEmail = namespaces.profile.email.value
+
+  const metricsData = await fetchMetricsData(
+    accountName,
+    metricsParam.workspace,
+    metricsParam.quoteId,
+    userEmail
+  )
+
+  const metric: QuoteMetric = {
+    name: 'b2b-suite-buyerorg-data',
+    kind: 'create-quote-ui-event',
+    description: 'Create Quotation Action - UI',
+    account: accountName,
+    fields: {
+      buy_org_id: metricsData.organization,
+      buy_org_name: metricsData.organizationName,
+      cost_center_id: metricsData.costId,
+      cost_center_name: metricsData.costCenterName,
+      member_id: namespaces?.profile?.id?.value,
```

Comment:
In line 98 the code is assuming that `namespaces.profile.email.value` is always set, but here you're handling the case where `namespaces` and `profile` might not be set. If that's a possibility we need to also add this to lines 97 and 98, skipping the whole metric process in case the values are not set (since those are required).

Reply:
Instead of skipping the while metric process, I'll check the namespace, profile, and email, and let it save null values. I believe it's better to save the metric with the data we have than not save it at all

Reply:
I agree, if the metric is still useful, let's save it!